### ISSUE TITLE
Memory leak happening while using registerModule/unregisterModule.

### ISSUE THREAD

**Issue by skkart** [vuejs/vuex#1507]
### Version
3.1.0

### Reproduction link
[https://jsfiddle.net/mrj8spu0/](https://jsfiddle.net/mrj8spu0/)

### Steps to reproduce

1. Open the JSFiddle link mentioned as part of this issue.
2. Check and record the heap memory in the browser on initial stage.
3. Click on the "Register" button to dynamically register Module B 100 times
4. Now we will see an increase in heap memory on account of registering module 100 times.
5. Click on "Un-Register" button to dynamically un-register Module B 100 times
6. Check the heap memory again, at this point will see an jump in memory usage on account of un-registering 100 module. 

### What is expected?
Heap memory generated on step 3 & 5 must be cleared after step 6. (Final stage)

### What is actually happening?
Heap memory of the browser is not getting cleared.

---
We are observing memory leak of Store._vm while using registerModule/unregisterModule extensively.
On heap profiling we found that the oldVm instance is not getting garbage collected.

Probable Solution:
De-reference the oldVm instance on the same function scope where its getting instantiated (i.e. resetStoreVM() )

The above can be achieved by adding following code inside resetStoreVM function.

oldVm.$destroy(); 
oldVm = null; // This is the missing code

Attached google chrome (V 72.0.3626.96) heap snapshot and file.
![heapsnapshot](https://user-images.githubusercontent.com/38841241/52727462-4ac0bf80-2fdb-11e9-91ce-7dd012cba4b5.png)

[Heap-20190213T220212.heaptimeline.zip](https://github.com/vuejs/vuex/files/2861246/Heap-20190213T220212.heaptimeline.zip)

**Comment by parth67** [comment]
Closure function at https://github.com/vuejs/vuex/blob/dev/src/store.js#L259 is retaining environment which contains oldVm reference.

Check screenshot GC root is able to reach oldVm object
![screenshot from 2019-02-14 10-50-29](https://user-images.githubusercontent.com/3313420/52764573-bdb35000-3046-11e9-9b07-3a33363d433f.png)

**Comment by tehnorm** [comment]
Possibly related. We are seeing a memory leak in a mutation:

```
[types.RECORD_MEASUREMENT](st, measurement) {
  Vue.set(st.lastMeasurementByDevice, measurement.device_id, measurement);
},
```
This seems to help the leak. 
```
[types.RECORD_MEASUREMENT](st, measurement) {
  Vue.set(st.lastMeasurementByDevice, measurement.device_id, Object.freeze(measurement));
},
```

Still working on a succinct reproducible case.

**Comment by parth67** [comment]
@tehnorm can you share sample state declaration and sample measurement object?

**Comment by wolfgangwalther** [comment]
The fix in pr#1508 doesn't change anything for me, but the original suggestion by @skkart (adding `oldVm = null`) does solve the memory leak for me.

**Comment by afwn90cj93201nixr2e1re** [comment]
@ktsn what about the last one comment?

**Comment by wolfgangwalther** [comment]
finally resolved by #1546
